Write some overview docs (#31)
pyhandle/hpy: docs/overview.rst

``simple.c`` on CPython 3.7, you get a DLL called
``simple.cpython-37m-x86_64-linux-gnu.so``:

  - ``cpython-37m`` is the ABI tag, in this case CPython 3.7

  - ``x86_64`` is the CPU architecture

  - ``linux-gnu`` is the operating system

The same source code compiled on PyPy3.6 7.2.0 results in a file called
``simple.pypy3-72-x86_64-linux-gnu.so``:

  - ``pypy3-72`` is the ABI tag, in this case "PyPy3.x", version "7.2.x"

The HPy C API is exposed to the user by including ``hpy.h`` and it is
explained in its own section of the documentation.

.. _hpy-target-abis:

Target ABIs
----------------

Depending on the compilation options, and HPy extension can target three
different ABIs:

CPython
  In this mode, HPy is implemented as a set of C macros and ``static inline``
  functions which translate the HPy API into the CPython API at compile
  time. The result is a compiled extension which is indistinguishable from a
  "normal" one and can be distributed using all the standard tools and will
  run at the very same speed. The ABI tag is defined by the version of CPython
  which is used to compile it (e.g., ``cpython-37m``),

HPy Universal

>   As the name suggests, the HPy Universal ABI is designed to be loaded and

Reviewer: Worth mentioning that such packages can be distributed as wheels using the [PEP 384] stable ABI tag `abi3` (see also [the PEP 384 section of PEP 3149](https://www.python.org/dev/peps/pep-3149/#pep-384)? In any case, does this part need a PEP or would this be an import hook that adds that so signature? 

Also: the HPy Hybrid mode would force the "regular" ABI tag `cpython-37m` for the packaged wheels.

Maybe worth adding a short "from the user's perspective" section that would explain this: the CPython mode would not change anything, the HPyUniversal and HPy Hybrid would require an additional install on CPython.

Author: > Worth mentioning that such packages can be distributed as wheels using the [PEP 384] stable ABI tag abi3 (see also the PEP 384 section of PEP 3149? In any case, does this part need a PEP or would this be an import hook that adds that so signature?

Actually, at the moment `hpy.universal` on CPython does NOT use the limited API, e.g. we use `PyTuple_GET_ITEM` in `ctx_CallRealFunctionFromTrampoline`. We would need to run benchmarks and so how much it costs to use the limited ABI.

> Maybe worth adding a short "from the user's perspective" section that would explain this: the CPython mode would not change anything, the HPyUniversal and HPy Hybrid would require an additional install on CPython.

good idea

Author: @mattip I'm going to merge this PR. We can always improve this paragraph later, but at least if we merge this we put online some docs

-----------------

If you use Cython, you can't use HPy directly. The plan is to write a Cython
backend which emits HPy code instead of Python/C code: once this is done, you
will get the benefits of HPy automatically.

Extensions in other languages
------------------------------

On the API side, HPy is designed with C in mind, so it is not directly useful
if you want to write an extension in a language different than C.

However, Python bindings for other languages could decide to target the HPy
Universal ABI instead of the CPython ABI, and generate extensions which can be
loaded seamlessly on all Python implementations which supports it.  This is
the route taken for example by Rust (XXX put a link?).

Benefits for alternative Python implementations
------------------------------------------------

If you are writing an alternative Python implementation, there is a good
chance that you already know how painful it is to support the Python/C
API. HPy is designed to be both faster and easier to implement!

You have two choices:

  - support the Universal ABI: in this case, you just need to export the
    needed functions and to add a hook to ``dlopen()`` the desired libraries

>   - use a custom ABI: in this case, you have to write your own replacement for

Reviewer 2: ```suggestion
  - use a custom ABI: in this case, you have to write your own replacement for
  + use a custom ABI: in this case, you have to write your own replacement for JUST A TEST
```

Current status and roadmap
--------------------------

At the moment of writing, HPy is still in its early stages of development. The
following milestones have been reached:

  - it is possible to write extensions which expose module-level functions,
    with all the various kinds of calling conventions

  - there is a limited support for argument parsing (only a couple of basic
    types actually work)

  - there is support for raising and catching exceptions

  - it is possible to choose between the CPython and the HPy Universal ABIs
    when compiling an extension module

  - extensions compiled with the CPython ABI work out of the box on CPython

  - it is possible to load HPy Universal extensions on CPython, thanks to the
    ``hpy.universal`` package

  - it is possible to load HPy Universal extensions on
    PyPy (using the PyPy `hpy branch <https://foss.heptapod.net/pypy/pypy/tree/branch/hpy>`_

  - it is possible to load HPy Universal extensions on GraalPython (XXX link)

However, there is still a long road before HPy is usable for the general
public. In particular, the following features are on our roadmap but has not
been implemented yet:

>   - it is not possible to write custome types in C. There is already a WIP

Reviewer: Typo and expanding the idea
```suggestion
  - it is not possible to write custom types (like NumPy's ndarray) in C. There is already a WIP
```

    branch to address this issue

>   - only a handful of the original Python/C functions has been ported to

Reviewer: ```suggestion
  - only a handful of the original Python/C functions have been ported to
```

    HPy. Porting most of them is straighforward, so for now the priority is to
    work on the "hard" features to prove that the HPy approach works, and we
    port new functions as needed

>   - the debug mode simply does not exist (yet!)

Reviewer: Link to a discussion of how to implement? Otherwise it sounds like a pie-in-the-sky idea that may or may not work

Author: We don't have any discussion in a written form AFAIK.
I originally plan to write also a document `debugging.rst` which explains some of the ideas, but I thought it is better to start merging this at first, to allow smaller reviews

  - there is not a standard/easy way to integrate HPy in a
    ``distutils/setuptools`` based workflow. The only way to compile HPy

>     extensions right now is to manually close the git repo and tweak your

Reviewer: ```suggestion
    extensions right now is to manually clone the git repo and tweak your
```

    ``setup.py``. Eventually, we plan to offer a workflow which integrates

>     seamlessly with ``pip``, ``setuptools``, etc.

Reviewer: Worth thinking about [PEP 517](https://www.python.org/dev/peps/pep-0517/) type of build environment (not for this doc, but something to think about)

Author: I didn't know PEP 517. Yes, it's worth thinking about it

    of course, but if the overhead of the HPy Universal ABI is only 10% on
    CPython, for many projects it might be small enough to avoid caring about
    compiling/distributing extensions with the CPython ABI, in the long term.

  - On PyPy, ``ujson-hpy`` runs 3x faster than the original ``ujson``. Note
    the HPy implementation on PyPy is not fully optimized yet, so we expect
    even bigger speedups eventually.

Projects involved
-----------------

HPy was born during EuroPython 2019, were a small group of people started to
discuss about the problems of the Python/C API and how it would be nice to
have a way to fix them.  Since then, it gathered the attention and interest of
people which are involved in many projects of the Python ecosystem.  The
following is a (probably incomplete) list of projects whose core developers
are involved in HPy, in one way or the other.  The mere presence in this list
does not mean that the project as a whole endorse or recognize HPy in any way,
just that some of the people involved contributed to the
code/design/discussions of HPy:

  - PyPy

  - CPython

  - Cython

>   - Numpy (XXX for mattip: do you think it is good/correct to include numpy here?)

Reviewer: Probably not. I didn't discuss it on any public mailing list nor with the NumPy steering council.

Author: fixed by 4cbf1ee
